File: 1856-maximum-subarray-min-product/1856-maximum-subarray-min-product.py
class Solution:
  def maxSumMinProduct(self, nums: List[int]) -> int:
    n = len(nums)
    st = []
    left = [0] * n
    right = [0] * n
    prefix_sum = [0] * (n + 1)

    for i in range(0, n, 1):
      while st and nums[i] <= nums[st[-1]]:
        st.pop()

      if len(st) > 0:
        left[i] = st[-1] + 1
      else:
        left[i] = 0

      st.append(i)

      prefix_sum[i + 1] = prefix_sum[i] + nums[i]

    st = []
    for i in range(n - 1, -1, -1):
      while st and nums[i] <= nums[st[-1]]:
        st.pop()

      if len(st) > 0:
        right[i] = st[-1] - 1
      else:
        right[i] = n - 1

      st.append(i)

    res = 0
    for i in range(0, n, 1):
      sum = prefix_sum[right[i] + 1] - prefix_sum[left[i]]
      res = max(res, nums[i] * sum)

    return res % 1_000_000_007

File: 1856-maximum-subarray-min-product/1856-maximum-subarray-min-product.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSumMinProduct = function(nums) {
  let stack = [];
  const left = Array(nums.length).fill(0);
  const right = Array(nums.length).fill(0);
  const prefixSum = Array(nums.length + 1).fill(0);
  
  nums.forEach((num, i) => {
    while (stack.length > 0 && num <= nums[stack[stack.length - 1]]) {
      stack.pop();
    }
    
    if (stack.length > 0) {
      left[i] = stack[stack.length - 1] + 1;
    } else {
      left[i] = 0;
    }
    
    stack.push(i);
    
    prefixSum[i + 1] = prefixSum[i] + num;
  });
  
  stack = [];
  
  for (let i = nums.length - 1; i >= 0; i--) {
    const num = nums[i];

    while (stack.length > 0 && num <= nums[stack[stack.length - 1]]) {
      stack.pop();
    }
    
    if (stack.length > 0) {
      right[i] = stack[stack.length - 1] - 1;
    } else {
      right[i] = nums.length - 1;
    }
    
    stack.push(i);
  }
  
  let res = 0;
  nums.forEach((num, i) => {
    const sum = prefixSum[right[i] + 1] - prefixSum[left[i]]
    const curr = BigInt(num) * BigInt(sum);
    res = curr > res ? curr : res;
  });
  
  return BigInt(res) % BigInt(10 ** 9 + 7);
};